### PR TITLE
NotificationView Migrate to DAOMenu2 + GroupBy Support and other cleanups

### PR DESCRIPTION
**Pull request by tharmaman**
So a lot more work ended up having to be done

1. Previously NotificationView was its own custom controller, we moved this to use the standardized controller rendered by DAOMenu2.
2. Added a DateGroupingExpr to work with the GroupingDAOList.
3. Added ability to pass in a SummaryView to the DAOControllerConfig (may have to update the naming of this since a SummaryView individually entails the normal DAODetailView, etc. was thinking BrowserSummaryView makes more sense) 
4. Added ability to optionally hide the query bar on the DAOBrowserView to the DAOControllerConfig
5. Tried to generalize the context menu actions on NotificationRowView so that it doesnt just specifically work with NotificationView itself, not sure if it was even working as expected before
6. Added isAvailable to read and markUnread context menu actions
7. Fixed spacing on the DAOTable based on comments from Logan
8. Added MyNotificationDAO as decorator which handles showing the users OWN notifications if they have permission to view all notifications
9. Wired up the support from summaryView import so that it can update the summaryView if a notificatoin was marked as read or unread. Previously you had to refresh the page and currently with notificationView it just pops a new stack of the notificationView not sure if htis works but I made work around so it will use this if no browse
10. Added canned queries for read and unread notifications
11. Changed Read to => Mark as Read and Mark Unread => Mark as Unread after talking with Logan

### CODE REVIEW

File: src/foam/comics/v2/DAOControllerConfig.js [hunk 1]
```diff
@@ -61,6 +61,21 @@ foam.CLASS({
       name: 'browseSubtitle',
       expression: function(of) { return "View all " + of.model_.plural.toLowerCase() +'.' }
     },
+    {
+      class: 'foam.u2.ViewSpecWithJava',
```
Comment (tharmaman) [inline src/foam/comics/v2/DAOControllerConfig.js:66]: @kgrgreer RE: you comment on my closed pr (https://github.com/foam-framework/foam2/pull/3481) about changing this to ViewSpec and then making ViewSpecWithJava a refinement on ViewSpec. Was thinking I should inlcude the change after this PR is done in a separate PR so that this PR doesn't get too bogged down

File: src/foam/mlang/expr/DateTypes.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+foam.ENUM({
+  package: 'foam.mlang.expr',
+  name: 'DateTypes',
+  documentation: `
+      An enum of date grouping types
+  `,
+
+  values: [
+      { 
+        name: 'HOUR',
```
Comment (jlhughes) [inline src/foam/mlang/expr/DateTypes.js]: should this be HOURS to be consistent?

File: src/foam/nanos/notification/NotificationRowView.js [hunk 3]
```diff
@@ -90,38 +98,128 @@
...
+          //     currentFeedback = currentFeedback.next;
+          //   }
+          // } else {
+          //   self.ctrl.add(self.NotificationMessage.create({
+          //     message: e.message,
+          //     type: 'error'
+          //   }));
+          // }
+
+          if ( e.message === 'An approval request has been sent out.' ) {
```
Comment (jlhughes) [inline src/foam/nanos/notification/NotificationRowView.js:162]: Is there anyway to use the message from the Approval process - this seems brittle.  And will the message change with locale?

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+          if ( 
+            objDiffFromTodayConverted >= group.low && 
+            objDiffFromTodayConverted < group.high 
+          ) groupName = group.name;
+        })
+
+        return groupName;
+      },
+      javaCode: `
+        return "test";
```
Comment (nanoscott) [inline src/foam/mlang/expr/DateGroupingExpr.js]: test?

File: src/foam/nanos/notification/MyNotificationDAO.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  documentation: `
+    DAO decorator that adds a filter to retrieve only the context subject user's notifications
+  `,
+
+
+  methods: [
+    {
+      name: 'select_',
+      javaCode: `
+      User user = ((Subject) x.get("subject")).getUser();
```
Comment (nanoscott) [inline src/foam/nanos/notification/MyNotificationDAO.js]:
```suggestion
        User user = ((Subject) x.get("subject")).getUser();
```

Comment (nanoscott) [inline src/foam/nanos/notification/MyNotificationDAO.js]: you want this to be the RealUser, since businesses don't get notifications.

File: src/foam/nanos/notification/NotificationRowView.js (hunk 3 above)
Comment (tharmaman) [reply src/foam/nanos/notification/NotificationRowView.js:162]: So this is temporary and supposed to be replaced by Nick's PR for the UserFeedbackException but it hasn't been merged in yet.

File: src/foam/mlang/expr/DateGroupingExpr.js (hunk 4 above)
Comment (tharmaman) [reply src/foam/mlang/expr/DateGroupingExpr.js]: Ill just remove the javaCode implementation for now

File: src/foam/mlang/expr/DateTypes.js (hunk 2 above)
Comment (tharmaman) [reply src/foam/mlang/expr/DateTypes.js]: Done.

File: src/foam/nanos/notification/MyNotificationDAO.js (hunk 5 above)
Comment (tharmaman) [reply src/foam/nanos/notification/MyNotificationDAO.js]: Done.

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'DateGroupingExpr',
+  extends: 'foam.mlang.AbstractExpr',
+
+  implements: [ 'foam.core.Serializable' ],
+
+  documentation: `
+    An expr whose value is the name of a date grouping.
+    Currently works with hours, days, weeks, months, years/
```
Comment (nanoscott) [inline src/foam/mlang/expr/DateGroupingExpr.js]:
```suggestion
    Currently works with hours, days, weeks, months, and years.
```

File: src/foam/nanos/notification/MyNotificationDAO.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  javaImports: [
+    'foam.core.FObject',
+    'foam.dao.*',
+    'foam.mlang.MLang',
+    'foam.mlang.predicate.AbstractPredicate',
+    'foam.mlang.predicate.Predicate',
+    'foam.nanos.auth.User',
+    'foam.nanos.auth.Subject',
+    'foam.core.X',
```
Comment (nanoscott) [inline src/foam/nanos/notification/MyNotificationDAO.js]: sort imports.

**Review by nanoscott — APPROVED** [review]
A couple of minor comments.

**Review by jlhughes — APPROVED** [review]

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 1 above)
Comment (kgrgreer) [reply src/foam/comics/v2/DAOControllerConfig.js:66]: Okay, then at least add a TODO.

**Comment by kgrgreer** [comment]
Still has two unresolved comments. Please address and I'll merge.

File: src/foam/comics/v2/DAOControllerConfig.js (hunk 1 above)
Comment (tharmaman) [reply src/foam/comics/v2/DAOControllerConfig.js:66]: Done.

File: src/foam/nanos/notification/MyNotificationDAO.js (hunk 7 above)
Comment (tharmaman) [reply src/foam/nanos/notification/MyNotificationDAO.js]: done
